src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1,192 +1,256 @@
+///
+/// Note that when the required constructor does not match the constructor on top of the stack, we
+/// return nothing. Thus specialization filters pattern-stacks by the constructor on top of them.
+///
+/// We call a list of pattern-stacks a "matrix", because in the run of the algorithm they will
+/// keep a rectangular shape. `S` operation extends straightforwardly to matrices by
+/// working row-by-row using flat_map.
+///
+///
+/// # Abstract algorithm
+///
+/// The algorithm itself is a function `U`, that takes as arguments a matrix `M` and a new pattern
+/// `p`, both with the same number `n` of columns.
+/// The algorithm is inductive (on the number of columns: i.e., components of pattern-stacks).
+/// The algorithm is realised in the `is_useful` function.
+///
+/// Base case. (`n = 0`, i.e., an empty tuple pattern)
+///     - If `M` already contains an empty pattern (i.e., if the number of patterns `m > 0`),
+///       then `U(M, p)` is false.
+///     - Otherwise, `M` must be empty, so `U(M, p)` is true.
+///
+/// Inductive step. (`n > 0`)
+///     We look at `p_1`, the head of the pattern-stack `p`.
+///
+///     We first generate the list of constructors that are covered by a pattern `pat`. We name
+///     this operation `pat_constructors`.
+///         - If `pat == c(r_1, .., r_a)`, i.e. we have a constructor pattern. Then we just
+///         return `c`:
+///             `pat_constructors(pat) = [c]`
+///
+///         - If `pat == _`, then we return the list of all possible constructors for the
+///         relevant type:
+///             `pat_constructors(pat) = all_constructors(pat.ty)`
+///
+///         - If `pat == r_1 | r_2`, then we return the constructors for either branch of the
+///         OR-pattern:
+///             `pat_constructors(pat) = pat_constructors(r_1) + pat_constructors(r_2)`
 ///
-/// For example, say we have the following:
+///     Then for each constructor `c` in `pat_constructors(p_1)`, we want to check whether a value
+///     that starts with this constructor may show that `p` is useful, i.e. may match `p` but not
+///     be matched by the matrix above.
+///     For that, we only care about those rows of `M` whose first component covers the
+///     constructor `c`; and for those rows that do, we want to unpack the arguments to `c` to check
+///     further that `p` matches additional values.
+///     This is where specialization comes in: this check amounts to computing `U(S(c, M), S(c,
+///     p))`. More details can be found in the paper.
+///
+///     Thus we get: `U(M, p) := ∃(c ϵ pat_constructors(p_1)) U(S(c, M), S(c, p))`
+///
+///     Note that for c ϵ pat_constructors(p_1), `S(c, P)` always returns exactly one element, so
+///     the formula above makes sense.
+///
+/// This algorithm however has a lot of practical issues. Most importantly, it may not terminate
+/// in the presence of recursive types, since we always unpack all constructors as much
```

Comment:
> it may not terminate in the presence of recursive types

By that, you mean "this algorithm as written does not know how to pop wildcards, and therefore visits all the type when matching against a wildcard"?

Reply:
Absolutely. I'm deliberately avoiding mention of popping wildcards in this version of the algorithm since I don't want wildcards to be a special case anymore.